File: demo24/src/demo/src/new_turtle.cpp
/*
    生成一只新的小乌龟
*/

#include "ros/ros.h"
#include "ros/service_client.h"
#include "turtlesim/Spawn.h"

int main(int argc, char *argv[]){
    // 1.初始化节点
    ros::init(argc, argv, "new_turtle");
    // 2.获得句柄
    ros::NodeHandle hd;

    // 3.创建一个乌龟的spawn请求信息
    turtlesim::Spawn spawn;
    spawn.request.name = "turtle2";
    spawn.request.x = 1.0;
    spawn.request.y = 2.0;
    spawn.request.theta = 0.5;

    // 4.创建一个发送spawn信息的客户端
    ros::ServiceClient client = hd.serviceClient<turtlesim::Spawn>("/spawn");
    // 5.发送信息
    client.waitForExistence();
    bool flag = client.call(spawn);
    if (flag){
        ROS_INFO("Spawn Successed!");
    }else{
        ROS_ERROR("Error: Spawn Faild");
    }

    return 0;
}

File: demo24/src/demo/scripts/new_turtle.py
from http import client
import rospy
from turtlesim.srv import SpawnRequest, Spawn


if __name__ == "__main__":
    # 1.启动节点
    rospy.init_node("turtle2")
    # 2.获得一个发布者
    pub = rospy.ServiceProxy("/spawn", Spawn)

    # 3.准备一个发布消息
    spawn = SpawnRequest()
    spawn.name = "turtle2"
    spawn.x = 1.0
    spawn.y = 2.0
    spawn.theta = 0.0
    
    pub.wait_for_service()
    try:
        response = pub.call(spawn)
        rospy.loginfo("Spawn successed! turtlename is:%s", response.name)
    except Exception as e:
        rospy.logerr("Error, Msgs")    